src/v/cloud_topics/dl_stm/dl_stm_api.cc

```diff
@@ -68,4 +70,81 @@ std::optional<dl_overlay> dl_stm_api::lower_bound(kafka::offset offset) const {
     return _stm->_state.lower_bound(offset);
 }
 
+ss::future<checked<dl_snapshot_id, dl_stm_api_errc>>
+dl_stm_api::start_snapshot() {
+    model::term_id term = _stm->_raft->term();
+
+    vlog(_logger.debug, "Replicating dl_stm_cmd::start_snapshot_cmd");
+
+    storage::record_batch_builder builder(
+      model::record_batch_type::dl_stm_command, model::offset(0));
+    builder.add_raw_kv(
+      serde::to_iobuf(dl_stm_key::start_snapshot),
+      serde::to_iobuf(start_snapshot_cmd()));
+
+    auto batch = std::move(builder).build();
+    auto reader = model::make_memory_record_batch_reader(std::move(batch));
+
+    auto opts = raft::replicate_options(raft::consistency_level::quorum_ack);
+    opts.set_force_flush();
+    auto res = co_await _stm->_raft->replicate(term, std::move(reader), opts);
+
+    if (res.has_error()) {
+        throw std::runtime_error(
+          fmt::format("Failed to replicate snapshot: {}", res.error()));
+    }
+
+    // We abuse knowledge of implementation detail here to construct the
+    // dl_snapshot_id without having to setup listeners and notifiers of command
+    // apply.
+    auto expected_id = dl_snapshot_id(dl_version(res.value().last_offset));
+
+    auto applied = co_await _stm->wait_no_throw(
+      res.value().last_offset, model::timeout_clock::now() + 30s);
+    if (!applied) {
+        co_return outcome::failure(dl_stm_api_errc::timeout);
+    }
+
+    // Ensure that the expected snapshot was created.
+    vassert(_stm->_state.snapshot_exists(expected_id), "Snapshot not found");
+
+    co_return outcome::success(expected_id);
+}
+
+std::optional<dl_snapshot_payload>
+dl_stm_api::read_snapshot(dl_snapshot_id id) {
+    return _stm->_state.read_snapshot(id);
+}
+
+ss::future<checked<void, dl_stm_api_errc>>
+dl_stm_api::remove_snapshots_before(dl_version last_version_to_keep) {
+    model::term_id term = _stm->_raft->term();
+
+    vlog(_logger.debug, "Replicating dl_stm_cmd::remove_snapshots_cmd");
+
+    storage::record_batch_builder builder(
+      model::record_batch_type::dl_stm_command, model::offset(0));
+    builder.add_raw_kv(
+      serde::to_iobuf(dl_stm_key::remove_snapshots_before_version),
+      serde::to_iobuf(
+        remove_snapshots_before_version_cmd(last_version_to_keep)));
+
+    auto batch = std::move(builder).build();
+    auto reader = model::make_memory_record_batch_reader(std::move(batch));
+
+    auto opts = raft::replicate_options(raft::consistency_level::quorum_ack);
+    opts.set_force_flush();
+    auto res = co_await _stm->_raft->replicate(term, std::move(reader), opts);
+
+    if (res.has_error()) {
+        throw std::runtime_error(
+          fmt::format("Failed to replicate remove snapshots: {}", res.error()));
+    }
+
+    co_await _stm->wait_no_throw(
```

Comment:
this workflow (replicate and then wait until the command is applied) can be extracted and reused

Reply:
https://github.com/redpanda-data/redpanda/compare/a3c602abc74790097f357792c4a47a788c907131..b711212ebb1c90543011638370d980dee274b4e0